docs rework (#1264)
hyperledger-labs/fabric-token-sdk: docs/services.md

- `services/identity`: This is the identity service.
  It is responsible for the management of wallets, the long-term identities (X509 and Idemix), and the relative stores.
  It is used to back part of the Token and Driver API.
- `services/ttx`: This is the `token transaction service`.
  It helps the developers in assembling token requests and ultimately transactions for the backend.
  This package is backend agnostic and relies on the `network service` for backend specific operations.
- `services/network`: This is `network service`.
  It offers a unified way to handle different type of networks or backends (e.g. Fabric, and so on...).
  It abstracts away the complexity of the specific underlying network.
  It is used solely by other services. The drivers should not refer in any way to details of the backend to remain agnostic.
- `services/storage`: This is `storage service`.
  It contains the storage places needed by the Token-SDK.   
  It is used to back part of the Token and Driver API.
- `services/selector`: This is `token selector service`.
  It is used to select tokens in such way the risk of accidental double-spending is minimized.

>   Different implementations provides different trades-off.

Reviewer: ```suggestion
  Different implementations provide different trades-off.
```

>   Different implementations provides different trades-off.

Reviewer: What type of trades-off exist. Can we provide an example?

  It is used to back part of the Token and Driver API.

## The `services/network` service

The `network` service is responsible to provide to the other services a uniform and predictable interface to the backend (e.g. Fabric).
Internal the network service is structured similarly to the Token API.
There is a provider of network instances (`Provider`) and network instances (`Network`).

Here is pictorial representation of the network service:

![network_service.png](imgs/network_service.png)

The Fabric-based network relays on the Fabric Smart Client for its configuration and operations like: Query chaincodes, broadcast transactions, and so on.

At bootstrap time, the Token-SDK goes through the TMS defined in the configuration.
For each TMS, the network provider is asked to return the network instance identified by the `network` and `channel` value of the TMSID.
If the network provider fails, then the boostrap of the Token-SDK stack fails.

> Otherwise, the `Connect` function on the network instance `Network` on input the desired namespace (taken also from the TMS ID).

Reviewer: What happens when there is a bad namespace?

To instantiate a transaction from scratch, the developer needs:
- The TMS ID (network, channel, namespace) of the TMS for which the transaction needs to be created.
  This means that the system is configured with that TMS and the relative Fabric network is also configured.
  If the TMS ID is not specified, the TMSP returns the default TMS, if available.
  If a TMS instance has not been created yet, it gets created on the fly.
- The `Network` instance from the network service.
  This means that the system is configured with a backend that can be localed using the TMS ID.
- If the transaction is anonymous, the network service is asked to return an anonymous identity.
  This identity is just the public part that will appear in the Fabric transaction.
  This anonymous identity is generated by the FSC stack.
  Moreover, the FSC `sig service` can produce valid signatures of any message for this identity.
  The constructor of the transaction then generates:
- A Transaction ID that can be used for both a Fabric transaction and a Token Request as anchor.
- `NewRequest` is invoked on the TMS that returns an instance of an empty token request.

When a transaction is created,

> it adds a hock to the FSC view context in such a way that if an error occurs all the resource used by the transaction are released (e.g. locked tokens).

Reviewer: ```suggestion
it adds a hook to the FSC view context in such a way that if an error occurs all the resources used by the transaction are released (e.g. locked tokens).
```

Can we provide other resources examples?

- Store the `transaction records` corresponding to the current version of the token request.
  This operation trigger a series of DB operations that will update the `Transactions DB` and `Tokens DB`.
  Moreover, a finality listener is added to the network service corresponding to the transaction's TMS ID (via a call to `AddFinalityListener`).
  This signals the network to alert when a notification comes about the status of the transaction.
  When the notification comes, the status of the row corresponding to the current token request is updated to `Valid` or `Deleted`.
- (distribution step) For each recipient, the owner of a token, the view opens a session to the node that owns that identity.
  The view send to this node the serialized version of the transaction stripping out from the metadata information that do not belong to that identity.
  Indeed, image that Alice sends token to Bob and Charlie at the same time. Bob doesn't need to know which tokens Alice is using nor that Charlie has also received something.

The `AcceptView` is invoked inside a responding view for a given business process.

> It is used evey time a party receives tokens.

Reviewer: ```suggestion
It is used every time a party receives tokens.
```

When invoking this view, it is expected to invoke first `ReceiveTransactionView` to receive either a transaction of a `SignatureRequest` message.
This view does the following:

> - Respond to all the requests of signatures as defined in the token request. Don't forget that the token request must be inspected before calling the AcceptView.

Reviewer: Format consistency: "Don't forget" is an informal phrase.

- Store the `transaction records` corresponding to the current version of the token request.
  This operation trigger a series of DB operations that will update the `Transactions DB` and `Tokens DB`.
  Moreover, a finality listener is added to the network service corresponding to the transaction's TMS ID (via a call to `AddFinalityListener`).
  This signals the network to alert when a notification comes about the status of the transaction.
  When the notification comes, the status of the row corresponding to the current token request is updated to `Valid` or `Deleted`.
- Send back an acknowledgement by signing with the FSC default identity the serialized version of the transaction.

Recall that also a recipient (who does not need to sign the token request) will receive the transaction.
The `AcceptView` handles also this case by managing the case that no SignatureRequest messages are sent.

The `AuditApproveView` is invoked inside a responding view for a given auditing business process.
The auditor is supposed to invoke `ReceiveTransactionView` first to receive the transaction to audit.

> The auditor will perform all the required business checks and if no error occurs during the check,

Author: they are application specific.
